bus-mapping/src/evm/opcodes/sha3.rs

```diff
@@ -1,33 +1,294 @@
-use crate::circuit_input_builder::{CircuitInputStateRef, ExecStep};
-use crate::evm::opcodes::stackonlyop::StackOnlyOpcode;
-use crate::evm::Opcode;
-use crate::Error;
-use eth_types::GethExecStep;
+use crate::{
+    circuit_input_builder::{
+        CircuitInputStateRef, CopyDataType, CopyEvent, CopyStep, ExecStep, NumberOrHash,
+    },
+    operation::RW,
+    Error,
+};
+use eth_types::{GethExecStep, Word, U256};
+use ethers_core::utils::keccak256;
 
-#[derive(Debug, Copy, Clone)]
+use super::Opcode;
+
+#[derive(Clone, Copy, Debug)]
 pub(crate) struct Sha3;
 
 impl Opcode for Sha3 {
     fn gen_associated_ops(
         state: &mut CircuitInputStateRef,
         geth_steps: &[GethExecStep],
     ) -> Result<Vec<ExecStep>, Error> {
-        // TODO: add rw operations needed, like memory reads and other lookups
         let geth_step = &geth_steps[0];
+        let mut exec_step = state.new_step(geth_step)?;
+
+        let expected_sha3 = geth_steps[1].stack.last()?;
 
-        log::warn!("incomplete SHA3 implementation");
-        let steps = StackOnlyOpcode::<2, 1>::gen_associated_ops(state, geth_steps)?;
+        // byte offset in the memory.
+        let offset = geth_step.stack.nth_last(0)?;
+        state.stack_read(&mut exec_step, geth_step.stack.nth_last_filled(0), offset)?;
 
-        // reconstruction
-        let offset = geth_step.stack.nth_last(0)?.as_usize();
-        let length = geth_step.stack.nth_last(1)?.as_usize();
+        // byte size to read in the memory.
+        let size = geth_step.stack.nth_last(1)?;
+        state.stack_read(&mut exec_step, geth_step.stack.nth_last_filled(1), size)?;
 
-        if length != 0 {
+        if size.gt(&U256::zero()) {
             state
                 .call_ctx_mut()?
                 .memory
-                .extend_at_least(offset + length);
+                .extend_at_least(offset.as_usize() + size.as_usize());
+        }
+
+        let memory = geth_step
+            .memory
+            .read_chunk(offset.as_usize().into(), size.as_usize().into());
```

Comment:
we may disable memory trace later so step.memory will be empty( https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/614 ) .   https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/a86aa1d32d58dc73fc5d7d5670636f1cc0edfc92/bus-mapping/src/evm/opcodes/logs.rs#L143 memory can be read from call_ctx like this.